File: internal/handlers/matches.go
package handlers

import (
	"context"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/frc10101/TealTeam/internal/frc"
	"github.com/gin-gonic/gin"
)

// MatchDisplay represents match data prepared for display
type MatchDisplay struct {
	Description  string
	MatchNumber  int
	StartTime    time.Time
	TimeDisplay  string
	TimeStatus   string // "past", "current", "upcoming"
	RedTeams     []int
	BlueTeams    []int
	MinutesUntil int
}

// HandleMatchSchedule returns current matches in the 30-minute window
func (h *Handler) HandleMatchSchedule(c *gin.Context) {
	if !h.hasDB() {
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Database unavailable.",
		})
		return
	}

	user, _ := h.GetSessionUser(c)
	if user == nil {
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Sign in to view match schedule.",
		})
		return
	}

	session, err := h.GetSession(c)
	if err != nil || session.SelectedEventID == nil {
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Select an event to view live matches.",
		})
		return
	}

	eventID := *session.SelectedEventID

	// Get event details from database
	var event struct {
		ID        int
		TBAKey    *string
		Name      string
		StartDate *time.Time
	}

	if err := h.db.WithContext(c.Request.Context()).
		Table("events").
		Select("id, tba_key, name, start_date").
		Where("id = ?", eventID).
		Scan(&event).Error; err != nil {
		h.log.Error("failed to fetch event", "error", err)
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Unable to load selected event.",
		})
		return
	}

	if event.TBAKey == nil || *event.TBAKey == "" {
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Selected event is missing schedule data.",
		})
		return
	}

	// Extract event code from TBA key (format: {year}{event_code})
	eventCode := extractEventCode(*event.TBAKey)
	if eventCode == "" {
		h.log.Error("failed to extract event code from TBA key", "tba_key", *event.TBAKey)
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Unable to determine event code for schedule lookup.",
		})
		return
	}

	// Get current season
	season := 2026
	if seasonEnv := strings.TrimSpace(os.Getenv("FIRST_SEASON")); seasonEnv != "" {
		if parsed, err := strconv.Atoi(seasonEnv); err == nil {
			season = parsed
		}
	}

	// Fetch match schedule from FRC API
	username := strings.TrimSpace(os.Getenv("FIRST_API_USERNAME"))
	key := strings.TrimSpace(os.Getenv("FIRST_API_KEY"))
	if username == "" || key == "" {
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "FIRST API credentials are not configured.",
		})
		return
	}

	client := frc.NewClient(username, key)
	ctx, cancel := context.WithTimeout(c.Request.Context(), 15*time.Second)
	defer cancel()

	// Fetch qualification matches
	matches, err := client.GetMatchSchedule(ctx, season, eventCode, nil)
	if err != nil {
		h.log.Error("failed to fetch match schedule", "error", err, "event_code", eventCode)
		h.renderPartial(c, "match_schedule", map[string]any{
			"ScheduleMessage": "Could not fetch match schedule from FIRST API.",
		})
		return
	}

	// Filter matches within 15-minute window (past and future)
	now := time.Now()
	windowStart := now.Add(-15 * time.Minute)
	windowEnd := now.Add(15 * time.Minute)

	var displayMatches []MatchDisplay
	for _, match := range matches {
		startTime, err := time.Parse(time.RFC3339, match.StartTime)
		if err != nil {
			h.log.Warn("failed to parse match start time", "match", match.MatchNumber, "time", match.StartTime)
			continue
		}

		// Filter by time window
		if startTime.Before(windowStart) || startTime.After(windowEnd) {
			continue
		}

		// Separate teams by alliance
		var redTeams, blueTeams []int
		for _, team := range match.Teams {
			if strings.HasPrefix(team.Station, "Red") {
				redTeams = append(redTeams, team.TeamNumber)
			} else if strings.HasPrefix(team.Station, "Blue") {
				blueTeams = append(blueTeams, team.TeamNumber)
			}
		}

		// Determine time status
		timeStatus := "upcoming"
		minutesUntil := int(startTime.Sub(now).Minutes())
		if startTime.Before(now) {
			timeStatus = "past"
		} else if minutesUntil <= 5 {
			timeStatus = "current"
		}

		displayMatches = append(displayMatches, MatchDisplay{
			Description:  match.Description,
			MatchNumber:  match.MatchNumber,
			StartTime:    startTime,
			TimeDisplay:  startTime.Format("3:04 PM"),
			TimeStatus:   timeStatus,
			RedTeams:     redTeams,
			BlueTeams:    blueTeams,
			MinutesUntil: minutesUntil,
		})
	}

	data := map[string]any{
		"Matches":   displayMatches,
		"EventName": event.Name,
	}

	h.renderPartial(c, "match_schedule", data)
}

// extractEventCode extracts the event code from a TBA key
// TBA key format: {year}{event_code} (e.g., "2026mndu" -> "mndu")
func extractEventCode(tbaKey string) string {
	// Remove year prefix (first 4 digits)
	re := regexp.MustCompile(`^\d{4}(.+)$`)
	matches := re.FindStringSubmatch(tbaKey)
	if len(matches) > 1 {
		return strings.ToUpper(matches[1])
	}
	return ""
}

File: rust/tealteam-web/src/handlers/matches.rs
// Match schedule HTMX fragment (port of Controllers/MatchesController.cs).

use askama::Template;
use axum::extract::State;
use axum_extra::extract::cookie::CookieJar;
use chrono::{DateTime, Local, Utc};
use std::collections::HashMap;
use tracing::{error, warn};

use super::*;
use crate::connectivity;
use crate::first_api::FirstApiClient;
use crate::state::SharedState;

pub struct MatchDisplay {
    pub description: String,
    pub time_display: String,
    pub time_status: String,
    pub minutes_until: i64,
    pub red_teams: Vec<i32>,
    pub blue_teams: Vec<i32>,
}

#[derive(Template)]
#[template(path = "partials/match_schedule.html")]
pub struct MatchScheduleFragment {
    pub message: String,
    pub matches: Vec<MatchDisplay>,
}

pub async fn match_schedule(State(state): State<SharedState>, jar: CookieJar) -> HandlerResult {
    let msg = |m: &str| {
        render(&MatchScheduleFragment {
            message: m.to_string(),
            matches: Vec::new(),
        })
    };

    let Some(user) = current_user(&state.pool, &jar).await else {
        return Ok(msg("Sign in to view match schedule."));
    };
    let Some(session) = current_session(&state.pool, &jar).await else {
        return Ok(msg("Session error. Please refresh the page."));
    };
    let Some(event_id) = session.selected_event_id else {
        return Ok(msg("Select an event to view live matches."));
    };

    let event_row: Option<(String, Option<String>)> =
        sqlx::query_as("SELECT name, tba_key FROM events WHERE id = $1")
            .bind(event_id)
            .fetch_optional(&state.pool)
            .await
            .unwrap_or(None);
    let Some((event_name, tba_key)) = event_row else {
        return Ok(msg("Unable to load selected event."));
    };
    let Some(tba_key) = tba_key.filter(|k| !k.is_empty()) else {
        return Ok(msg("Selected event is missing schedule data."));
    };

    let event_code = extract_event_code(&tba_key);
    if event_code.is_empty() {
        error!("failed to extract event code from TBA key {tba_key}");
        return Ok(msg("Unable to determine event code for schedule lookup."));
    }

    let Some(client) = FirstApiClient::from_environment() else {
        return Ok(msg("FIRST API credentials are not configured."));
    };
    let season = FirstApiClient::season_from_environment();

    // FIRST schedule endpoint requires at least one filter parameter.
    let mut filters = HashMap::new();
    match user.team_number.filter(|n| *n > 0) {
        Some(n) => {
            filters.insert("teamNumber".to_string(), n.to_string());
        }
        None => {
            filters.insert("tournamentLevel".to_string(), "Qualification".to_string());
        }
    }

    let fetch =
        client.get_match_schedule(season, &event_code, &filters);
    let raw_matches = match tokio::time::timeout(std::time::Duration::from_secs(15), fetch).await {
        Ok(Ok(m)) => m,
        Ok(Err(e)) => {
            error!("failed to fetch match schedule for {event_code}: {e}");
            let stale = load_stale_matches(&state.pool, event_id).await;
            if !stale.is_empty() {
                let message = if connectivity::is_internet_unavailable(&e) {
                    "Offline: showing cached schedule from local DB."
                } else {
                    "Using cached schedule data from local DB."
                };
                return Ok(render(&MatchScheduleFragment {
                    message: message.to_string(),
                    matches: stale,
                }));
            }
            let message = if connectivity::is_internet_unavailable(&e) {
                "No internet connection available. Connect uplink and retry manual sync."
            } else {
                "Could not fetch match schedule from FIRST API."
            };
            return Ok(msg(message));
        }
        Err(_) => {
            let stale = load_stale_matches(&state.pool, event_id).await;
            if !stale.is_empty() {
                return Ok(render(&MatchScheduleFragment {
                    message: "Using cached schedule data from local DB.".to_string(),
                    matches: stale,
                }));
            }
            return Ok(msg("Could not fetch match schedule from FIRST API."));
        }
    };

    // Only show matches within a ±15 minute window.
    let now = Local::now();
    let window_start = now - chrono::Duration::minutes(15);
    let window_end = now + chrono::Duration::minutes(15);

    let mut display: Vec<(DateTime<Local>, MatchDisplay)> = Vec::new();
    for m in &raw_matches {
        let Ok(start_time) = DateTime::parse_from_rfc3339(&m.start_time) else {
            warn!("failed to parse match start time: match {} time {}", m.match_number, m.start_time);
            continue;
        };
        let start_time = start_time.with_timezone(&Local);
        if start_time < window_start || start_time > window_end {
            continue;
        }

        let mut red = Vec::new();
        let mut blue = Vec::new();
        for mt in &m.teams {
            if mt.station.starts_with("Red") {
                red.push(mt.team_number);
            } else if mt.station.starts_with("Blue") {
                blue.push(mt.team_number);
            }
        }

        let minutes_until = (start_time - now).num_minutes();
        let time_status = if minutes_until < -15 {
            "past"
        } else if minutes_until <= 5 {
            "current"
        } else {
            "upcoming"
        };

        display.push((
            start_time,
            MatchDisplay {
                description: m.description.clone(),
                time_display: start_time.format("%a %b %-d, %-I:%M %p").to_string(),
                time_status: time_status.to_string(),
                minutes_until,
                red_teams: red,
                blue_teams: blue,
            },
        ));
    }

    display.sort_by_key(|(t, _)| *t);
    let matches: Vec<MatchDisplay> = display.into_iter().map(|(_, m)| m).collect();

    let _ = event_name;
    Ok(render(&MatchScheduleFragment {
        message: String::new(),
        matches,
    }))
}

async fn load_stale_matches(pool: &sqlx::PgPool, event_id: i32) -> Vec<MatchDisplay> {
    let rows: Vec<(i32, Option<DateTime<Utc>>)> = sqlx::query_as(
        "SELECT match_number, scheduled_time FROM matches WHERE event_id = $1 ORDER BY scheduled_time",
    )
    .bind(event_id)
    .fetch_all(pool)
    .await
    .unwrap_or_default();

    let now = Utc::now();
    rows.into_iter()
        .map(|(match_number, scheduled_time)| {
            let mut item = MatchDisplay {
                description: format!("Match {match_number}"),
                time_display: String::new(),
                time_status: "upcoming".to_string(),
                minutes_until: 0,
                red_teams: Vec::new(),
                blue_teams: Vec::new(),
            };
            if let Some(scheduled) = scheduled_time {
                item.time_display = scheduled
                    .with_timezone(&Local)
                    .format("%a %b %-d, %-I:%M %p")
                    .to_string();
                let minutes_until = (scheduled - now).num_minutes();
                item.minutes_until = minutes_until;
                item.time_status = if minutes_until < -15 {
                    "past".to_string()
                } else if minutes_until <= 5 {
                    "current".to_string()
                } else {
                    "upcoming".to_string()
                };
            }
            item
        })
        .collect()
}
